pkg/volume/metadata/metadata.go

```diff
@@ -0,0 +1,255 @@
+	return &metadataVolume{volName: volName,
+		pod:     &api.Pod{ObjectMeta: api.ObjectMeta{UID: podUID}},
+		plugin:  plugin,
+		mounter: mounter}, nil
+}
+
+// metadataVolume handles retrieving metadata from the API server
+// and placing them into the volume on the host.
+type metadataVolume struct {
+	volName                string
+	fieldReferenceFileName map[string]string
+	pod                    *api.Pod
+	plugin                 *metadataPlugin
+	opts                   *volume.VolumeOptions
+	mounter                mount.Interface
+}
+
+// SetUp puts in place the volume plugin.
+// This function is not idempotent by design. We want the data to be refreshed periodically.
+func (m *metadataVolume) SetUp() error {
+	return m.SetUpAt(m.GetPath())
+}
+
+// This is the spec for the volume that this plugin wraps.
+var wrappedVolumeSpec = &volume.Spec{
+	Name:         "not-used",
+	VolumeSource: api.VolumeSource{EmptyDir: &api.EmptyDirVolumeSource{Medium: api.StorageTypeMemory}},
+}
+
+// createRandomSymlink try to create a temporary symbolic link
+func createRandomSymlink(newTarget string) (string, error) {
+	var runes = []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ")
+	b := make([]rune, 5)
+	for j := range b {
+		b[j] = runes[rand.Intn(len(runes))]
+	}
+	randomSymLink := string(b)
+	if e := os.Symlink(newTarget, randomSymLink); e == nil {
+		return randomSymLink, e
+	}
+	return "", fmt.Errorf("unable to create random symlink")
+}
+
+// readFile reads full path file name and return it as a string
+func readFile(fileName string) string {
+	if data, err := ioutil.ReadFile(fileName); err == nil {
+		return string(data)
+	}
+	return ""
+}
+
+// generateFile writes input parameter 'values' in a new file.
+// It creates a '.tmp.' prefixed temporary file and renames it afterwards
+func generateFile(values string) (string, error) {
+	const tmpPrefix = ".tmp."
+	prefix := tmpPrefix + time.Now().Format("01_01_010101")
+	tmpFile, err := ioutil.TempFile(".", prefix)
+	if err != nil {
+		return "", fmt.Errorf("unable to create tmp file '%s', with prefix '%s'", tmpFile.Name(), prefix)
+	}
+	if e := ioutil.WriteFile(tmpFile.Name(), []byte(values), 0644); e != nil {
+		return "", e
+	}
+	_, f := path.Split(tmpFile.Name())
+	newFileName := strings.TrimPrefix(f, tmpPrefix)
+	if e := os.Rename(tmpFile.Name(), newFileName); e != nil {
+		return "", e
+	}
+	return newFileName, nil
+}
+
+// isFileToGenerate compares actual file with the new values. If
+// different (or the file does not exist) return true,
+func isFileToGenerate(fileName, values string) bool {
+	return readFile(fileName) != values
+}
+
+// WriteFile writes string in fileName file if string must be updated.
+// It writes 'atomically' writing in a tmp file and re-placing the previous copy
+// via symbol linking recreation. It ends up removing the previous file
+func WriteFile(fileName, dirPath, values string) error {
```

Comment:
Also, suggest that you change the order of the args to `(dirPath, fileName, values)`